### PR TITLE
RISC-V: align with ARM, move rv32 CSR access wrappers

### PR DESCRIPTION
**Pull request by axel-h**
- move C code out of hardware.h

postponed for another PR
- always use the second value
- handle roll over from MAX to 0 properly (and halt on roll over)
- simplify expression to build a uint64_t
- fix comment to mention cycles and not timer

### CODE REVIEW

**Comment by axel-h** [comment]
Seems the URL check failure is a random internet hickup.

File: include/arch/riscv/arch/32/mode/hardware.h [hunk 1]
```diff
@@ -69,40 +69,38 @@
...
         "rdcycle  %1\n"
         "rdcycleh %2\n"
-        : "=r"(nH1), "=r"(nL), "=r"(nH2));
-    if (nH1 != nH2) {
+        : "=r"(nH_prev), "=r"(nL), "=r"(nH));
+    if (nH != nH_prev) {
         /* Ensure that the time is correct if there is a rollover in the
          * high bits between reading the low and high bits. */
         asm volatile(
             "rdcycle  %0\n"
```
Comment (axel-h) [inline include/arch/riscv/arch/32/mode/hardware.h:100]: Is there actually a strong reason to read the low value again and not just use `0` here? Rationale would be that there is some jitter on the low value anyway, but 0 would be closer to the actual point in time where we wanted to read things in the first place.

File: include/arch/riscv/arch/32/mode/hardware.h [hunk 2]
```diff
@@ -75,7 +75,7 @@ static inline uint64_t riscv_read_time(void)
         "rdtime  %1\n"
         "rdtimeh %2\n"
         : "=r"(nH1), "=r"(nL), "=r"(nH2));
-    if (nH1 < nH2) {
+    if (nH1 != nH2) {
```
Comment (kent-mcleod) [inline include/arch/riscv/arch/32/mode/hardware.h]: A 64-bit rollover of the kernel timer is not allowed to happen. It would be better to add a condition that halts if the nH1 > nH2 condition is seen.

File: include/arch/riscv/arch/32/mode/hardware.h (hunk 1 above)
Comment (kent-mcleod) [reply include/arch/riscv/arch/32/mode/hardware.h:100]: It seems like a reasonable optimization when you know that the value of `rdcycle`must have contained 0 at some time between the two reads of `rdcycleh`.

File: include/arch/riscv/arch/32/mode/hardware.h (hunk 2 above)
Comment (axel-h) [reply include/arch/riscv/arch/32/mode/hardware.h]: I wonder if this is something that should not be checked here, but at the higher layers in a generic way. The timer design for each platforms should be, that this does never roll over practically. This function just returns whatever is in the registers.

**Comment by axel-h** [comment]
I've added a check in `getCurrentTime()` for RISC-V now. But I wonder if we should have this for all platforms/architectures, as the same issue can also happen there. We could move `getCurrentTime()` out of the architecture code and make it do the check. Platforms/Architectures then just have to provide a function that retrieves the values, but don't have to reason much about it,

**Comment by axel-h** [comment]
I've factored out the timer sanity checks to https://github.com/seL4/seL4/pull/657

**Comment by axel-h** [comment]
Adding one commit that moves the CSR access code from `hardware.h` to `machine.h`, because having C code in `hardware.h` generated some include issues. It seem better to just have defines there and have the C code in `machine.h`

File: include/arch/riscv/arch/32/mode/machine.h [hunk 3]
```diff
@@ -8,4 +8,42 @@
...
+        "rdcycleh %0\n"
+        "rdcycle  %1\n"
+        "rdcycleh %2\n"
+        : "=r"(nH_prev), "=r"(nL), "=r"(nH)
+        : /* no input */
+        : "memory");
+    if (nH_prev != nH) {
+        /* Ensure that the time is correct if there is a rollover in the
+         * high bits between reading the low and high bits. */
+        asm volatile("rdtime  %0\n" : "=r"(nL) :: "memory");
```
Comment (kent-mcleod) [inline include/arch/riscv/arch/32/mode/machine.h:46]: This should be `rdcycle` instead of `rdtime`.

Comment (axel-h) [reply include/arch/riscv/arch/32/mode/machine.h]: thanks, fixed

File: include/arch/riscv/arch/32/mode/hardware.h [hunk 4]
```diff
@@ -62,46 +62,3 @@
 
 #define LOAD  lw
 #define STORE sw
-
```
Comment (pingerino) [inline include/arch/riscv/arch/32/mode/hardware.h:65]: this commit seems to shuffle code around with a new convention that we don't currently follow in the kernel?

File: include/arch/riscv/arch/32/mode/hardware.h [hunk 5]
```diff
@@ -69,40 +69,38 @@
 
 static inline uint64_t riscv_read_time(void)
 {
-    uint32_t nH1, nL, nH2;
+    uint32_t nH_prev, nH, nL;
```
Comment (pingerino) [inline include/arch/riscv/arch/32/mode/hardware.h]: what is the reasoning behind the changes in this commit?

**Review by pingerino — CHANGES_REQUESTED** [review]

File: include/arch/riscv/arch/32/mode/hardware.h (hunk 4 above)
Comment (axel-h) [reply include/arch/riscv/arch/32/mode/hardware.h:65]: Putting C code in hardware.h does not scale too well in my dev branches, especially when this starts pulling in more include dependencies. Thus I'm moving this to machine.h, there almost everything can be included without problems. Also, the assembler never sees this file, so it does no need defines.

File: include/arch/riscv/arch/32/mode/hardware.h (hunk 5 above)
Comment (axel-h) [reply include/arch/riscv/arch/32/mode/hardware.h]: The variable names are more intuitive. Can also keep the old names.

Comment (pingerino) [reply include/arch/riscv/arch/32/mode/hardware.h]: Changing the names makes the code new for anyone who understood it before. 1, 2 vs prev is not really intuitive, just a personal preference. To avoid code churn we should avoid renames like this.

File: include/arch/riscv/arch/32/mode/hardware.h (hunk 4 above)
Comment (pingerino) [reply include/arch/riscv/arch/32/mode/hardware.h:65]: Did you consider putting new C code in a different file,  and leaving the existing code as is?

Comment (axel-h) [reply include/arch/riscv/arch/32/mode/hardware.h:65]: Yes, but since the ARM port also has the special register access wrapper code in `machine.h` also, why not follow this on RISC-V also?

Comment (pingerino) [reply include/arch/riscv/arch/32/mode/hardware.h:65]: if it's following an existing convention, the change makes sense.

File: include/arch/riscv/arch/32/mode/machine.h [hunk 6]
```diff
@@ -6,6 +6,42 @@
...
+static inline uint64_t riscv_read_time(void)
+{
+    register word_t nH_prev, nH, nL;
+    asm volatile(
+        "rdtimeh %0\n"
+        "rdtime  %1\n"
+        "rdtimeh %2\n"
+        : "=r"(nH_prev), "=r"(nL), "=r"(nH)
+        : /* no input */
+        : "memory");
```
Comment (kent-mcleod) [inline include/arch/riscv/arch/32/mode/machine.h:22]: These "memory" clobbers shouldn't be inserted as part of the move

Comment (kent-mcleod) [reply include/arch/riscv/arch/32/mode/machine.h:22]: (They shouldn't be as the inline asm doesn't touch memory contents)

File: include/arch/riscv/arch/32/mode/machine.h [hunk 7]
```diff
@@ -6,6 +6,42 @@
...
+        "rdcycleh %0\n"
+        "rdcycle  %1\n"
+        "rdcycleh %2\n"
+        : "=r"(nH_prev), "=r"(nL), "=r"(nH)
+        : /* no input */
+        : "memory");
+    if (nH_prev != nH) {
+        /* Ensure that the cycles are correct if there is a rollover in the
+         * high bits between reading the low and high bits. */
+        asm volatile("rdcycle  %0\n" : "=r"(nL) :: "memory");
```
Comment (pingerino) [inline include/arch/riscv/arch/32/mode/machine.h:44]: this isn't a clean move - you've changed the code - can you please seperate the move into a different commit? it's nearly impossible to review if code is moved and changed in one commit

**Comment by axel-h** [comment]
I've removed all the commits that change a thing and just keep one commit here that moved the function from `hardware.h` to `machine.h`. This is the most fundamental thing to change. For the other changes I will create a new PR then.

File: include/arch/riscv/arch/32/mode/hardware.h (hunk 4 above)
Comment (freddy123098) [reply include/arch/riscv/arch/32/mode/hardware.h:65]: But if it does not like I am trying to see the changes made by this person but I can not see any changes made by him/her yet so now what should take this as?

**Review by freddy123098 — CHANGES_REQUESTED** [review]
requesting changes.

**Comment by freddy123098** [comment]
Proofs / C Proofs (ARM_HYP, CRefine) (pull_request_target) Failing after 14s — C Proofs (ARM_HYP, CRefine)

The AWS has failed and needs to be restarted. 

And the others that have failed also there AWS has to restarted. 


Thank you.

**Comment by axel-h** [comment]
Can we merge this one commit now to better align the RISCV and ARM port?

**Review by kent-mcleod — APPROVED** [review]
